### PR TITLE
Fixes #42

### LINKED ISSUE #42: Missing reset password link next to "Login with Nuvla Account"
Although the there is a "Rest password" item in the top bar, having it next to the password input field or button is more standard.

(and the reset password link in the top bar menu of the login page doesn't work)

### PR DESCRIPTION
**Pull request by mebster**
Add a new modal alongside login and sign-in for the internal authentication method only.
Add http-fx deps to handle direct http requests.
Consider re-implementing the reset feature on the /api (and remove http-fx?)

### CODE REVIEW

File: src/cljs/sixsq/slipstream/webui/authn/views.cljs [hunk 1]
```diff
@@ -180,10 +194,74 @@
...
+          [:p @success-message]])
+
+        [ui/FormInput {:name         "username"
+                      :type         "text"
+                      :placeholder  "Username"
+                      :icon         "user"
+                      :iconPosition "left"
+                      :required     true
+                      :autoComplete "off"
+                      :on-change #(dispatch [::authn-events/username (-> % .-target .-value)])}]
```
Comment (loomis) [inline src/cljs/sixsq/slipstream/webui/authn/views.cljs]: There's a callback utility to handle callback functions like this in `sixsq.slipstream.webui.utils.ui-callback` namespace.

Does the username need to be stored in the database at all?  Wouldn't a local atom work?

File: src/cljs/sixsq/slipstream/webui/authn/views.cljs [hunk 2]
```diff
@@ -180,10 +194,74 @@
...
           :login [:span (@tr [:no-account?]) " "
                   [:a {:on-click f :style {:cursor "pointer"}} (str (@tr [:signup-link]))]]
+          :reset-password [:span (@tr [:already-registered?]) " "
+                           [:a {:on-click f :style {:cursor "pointer"}} (str (@tr [:login-link]))]]
           :signup [:span (@tr [:already-registered?]) " "
                    [:a {:on-click f :style {:cursor "pointer"}} (str (@tr [:login-link]))]])))))
 
 
+(defn reset-password-modal
+  "Modal that holds the authentication (login or sign up) forms."
```
Comment (loomis) [inline src/cljs/sixsq/slipstream/webui/authn/views.cljs]: This comment doesn't seem relevant.

**Comment by mebster** [comment]
> When clicking on the "Reset pasword" button, there is no feedback that anything has happened.

A message is provided: success or error. What else do you have in mind?

> A "success" message is given even if the username is blank.

This is consistent with the current implementation. This feature is provided by the old API.  There's no support for password reset in the client.  And since the current implementation redirects, I couldn't get form submission to work.  This is why I'm using the http-fx library to perform an http post request.

As soon as the reset password feature is available in the new API and available in client, this direct http request (and the http-fx library) can go.  So should we delay the feature and implement password reset on the new API and client, or release this temporary fix and remove it later?

> If the username is stored in the database, then the form field should be initialized with the value. Opening the dialog a second time will send an email to the last username even if the field is blank.

You're right, it doesn't need to be stored.  I'll change that, thanks.

> I don't find the "Already registered?" link particularly appropriate in this dialog.

"Already have a login?" or "Already have an account?" better?

> There should be an explicit cancel button.

No other modes (login or signup) of this model have a cancel button.  Modals have a cross top right that acts as cancel and clicking outside the modal cancels as well.  I think this is correct and consistent. In this particular mode (reset password), the link next to the button allows users to navigate back to the login mode.

File: project.clj [hunk 3]
```diff
@@ -76,6 +76,7 @@
                  [reagent]
 
                  [re-frame]
+                 [day8.re-frame/http-fx "0.1.6"]
```
[the remark this answers is not in the document]
Comment (mebster) [reply project.clj]: See my comment below. What do you recommend we do? 

- [ ] move the dependency as you suggest?
- [ ] leave it here as a reminder that this is a hack
- [ ] delay feature until implemented in the new API

or something else?

[the remark this answers is not in the document]
Comment (mebster) [reply project.clj]: See comment above. Also, if you think there's another way to perform the post and handle the redirect, I'm all ears :-).

[the remark this answers is not in the document]
Comment (loomis) [reply project.clj]: Either of those is fine.

File: src/cljs/sixsq/slipstream/webui/authn/events.cljs [hunk 4]
```diff
@@ -1,11 +1,14 @@
 (ns sixsq.slipstream.webui.authn.events
   (:require
     [re-frame.core :refer [dispatch reg-event-db reg-event-fx]]
+    [ajax.core :as ajax]
```
[the remark this answers is not in the document]
Comment (mebster) [reply src/cljs/sixsq/slipstream/webui/authn/events.cljs]: Thanks, done.

**Comment by loomis** [comment]
> > When clicking on the "Reset pasword" button, there is no feedback that anything has happened.
> 
> A message is provided: success or error. What else do you have in mind?

It takes forever to get that feedback.  I ended up clicking twice.  Either an immediate message or a spinner.

> 
> > A "success" message is given even if the username is blank.
> 
> This is consistent with the current implementation. This feature is provided by the old API. There's no support for password reset in the client. And since the current implementation redirects, I couldn't get form submission to work. This is why I'm using the http-fx library to perform an http post request.

It is the same behavior as in the old UI, but allowing the user to send an obviously invalid request doesn't seem like a good choice to me. 
 
> > I don't find the "Already registered?" link particularly appropriate in this dialog.
> 
> "Already have a login?" or "Already have an account?" better?

It isn't the wording; I don't think having this at all makes sense.

File: src/cljs/sixsq/slipstream/webui/authn/views.cljs [hunk 5]
```diff
@@ -180,10 +194,74 @@
...
+          [:p @error-message]])
+
+        (when @success-message
+          [ui/Message {:negative  false
+                      :size      "tiny"
+                      :onDismiss #(dispatch [::authn-events/clear-success-message])}
+          [ui/MessageHeader (@tr [:reset-password-success])]
+          [:p @success-message]])
+
+        [ui/FormInput {:name         "username"
```
[the remark this answers is not in the document]
Comment (mebster) [reply src/cljs/sixsq/slipstream/webui/authn/views.cljs]: Doesn't work on textmate :-P. I need to switch to a real IDE.

**Comment by mebster** [comment]
> > > When clicking on the "Reset pasword" button, there is no feedback that anything has happened.
> > 
> > 
> > A message is provided: success or error. What else do you have in mind?
> 
> It takes forever to get that feedback. I ended up clicking twice. Either an immediate message or a spinner.

Interesting.  On my machine it's instantaneous.  So yes, a spinner or something is needed.

> 
> > > A "success" message is given even if the username is blank.
> > 
> > 
> > This is consistent with the current implementation. This feature is provided by the old API. There's no support for password reset in the client. And since the current implementation redirects, I couldn't get form submission to work. This is why I'm using the http-fx library to perform an http post request.
> 
> It is the same behavior as in the old UI, but allowing the user to send an obviously invalid request doesn't seem like a good choice to me.

I agree. But then we need to validate forms, which is not done for the other modes either.

> 
> > > I don't find the "Already registered?" link particularly appropriate in this dialog.
> > 
> > 
> > "Already have a login?" or "Already have an account?" better?
> 
> It isn't the wording; I don't think having this at all makes sense.

It was like this already. Having links that act as buttons for sub-workflow is standard. Here's an [example](https://subscribe.washingtonpost.com/loginregistration/index.html#/loginhome/group/long).

**Comment by loomis** [comment]
> > > > A "success" message is given even if the username is blank.
> > > 
> > > 
> > > This is consistent with the current implementation. This feature is provided by the old API. There's no support for password reset in the client. And since the current implementation redirects, I couldn't get form submission to work. This is why I'm using the http-fx library to perform an http post request.
> > 
> > 
> > It is the same behavior as in the old UI, but allowing the user to send an obviously invalid request doesn't seem like a good choice to me.
> 
> I agree. But then we need to validate forms, which is not done for the other modes either.

It's not a particularly difficult validation.  Putting `{:disabled (str/blank? @username)}` on the button would do it.

> > > > I don't find the "Already registered?" link particularly appropriate in this dialog.
> > > 
> > > 
> > > "Already have a login?" or "Already have an account?" better?
> > 
> > 
> > It isn't the wording; I don't think having this at all makes sense.
> 
> It was like this already. Having links that act as buttons for sub-workflow is standard. Here's an [example](https://subscribe.washingtonpost.com/loginregistration/index.html#/loginhome/group/long).

Between the login and sign up dialogs, the links make sense.  I don't think that it makes sense for the reset password dialog.  In the example you provide, if you follow through to the forgotten password page, there are no links back to the login or the sign up pages.

File: project.clj (hunk 3 above)
[the remark this answers is not in the document]
Comment (mebster) [reply project.clj]: I can't get my head around kvlt just now so going for 1. But I get a version error if I remove it from the project file. I must be doing something wrong such that it's not getting the version from the parent repo.

[the remark this answers is not in the document]
Comment (loomis) [reply project.clj]: You have to generate the SlipStreamParent package and set the `[sixsq/slipstream-parent "5.3.11"]` vector in the `project.clj` file to the new production/snapshot version of the parent.  To create a new production version of the parent, follow the README in SlipStreamParent.  All SlipStream packages will then have to be updated with the new version.

[the remark this answers is not in the document]
Comment (mebster) [reply project.clj]: Looks like a lot of work for a local and temporary lib.

**Comment by loomis** [comment]
@mebster The translations for the reset password phrases are still missing.  There is also still a "login" link next to the reset password, but don't know if you want to keep this consistent with the Washington Post workflow or not.

**Comment by mebster** [comment]
> @mebster The translations for the reset password phrases are still missing. There is also still a "login" link next to the reset password, but don't know if you want to keep this consistent with the Washington Post workflow or not.

The French translation requires a review, as I picked-up a few typos.  So I'd like to do it later.

The Washington Post example has a link back, but for some reason it's in the header of the dialogue. I think it's make sense to be able to come back to the login if a user clicks on the forget password by mistake.

**Pull request merged**